lib/storeToTriplestore.js

```diff
@@ -178,28 +179,39 @@ export async function insertData(store, graph) {
  */
 export async function deleteData(store, graph) {
   const deleteFunction = async (store, graph) => {
-    //const writer = new N3.Writer();
-    //store.forEach((q) => writer.addQuad(q.subject, q.predicate, q.object));
-    //const triplesSparql = await new Promise((resolve, reject) =>
-    //  writer.end((error, result) => {
-    //    if (error) reject(error);
-    //    else resolve(result);
-    //  })
-    //);
+    //Use a proper writer for when the data is properly formatted.
+    const writer = new N3.Writer();
+    store.forEach((q) => writer.addQuad(q.subject, q.predicate, q.object));
+    const triplesSparql1 = await new Promise((resolve, reject) =>
+      writer.end((error, result) => {
+        if (error) reject(error);
+        else resolve(result);
+      })
+    );
+    await mas.updateSudo(`
+      DELETE DATA {
+        GRAPH ${rst.termToString(graph)} {
+          ${triplesSparql1}
+        }
+      }`);
 
+    //Also use a self made writer to format the triples in a special way for
+    //Virtuoso. Only do this when dealing with explicit string typed terms. Do
+    //this in a separate query to deal with another Virtuoso bug.
     //Slightly less reliable and more verbose, but keeping the datatype is
     //necessary for Virtuoso to be able to delete the data. Another bug?
     const triplesSparql = [];
     store.forEach((quad) => {
-      triplesSparql.push(formatTriple(quad));
+      if (quad.object?.datatype?.value === ns.xsd`string`.value)
+        triplesSparql.push(formatTriple(quad));
     });
-
-    await mas.updateSudo(`
-      DELETE DATA {
-        GRAPH ${rst.termToString(graph)} {
-          ${triplesSparql.join('\n')}
-        }
-      }`);
+    if (triplesSparql.length)
+      await mas.updateSudo(`
```

Comment:
So... my  experience with virtuoso quirck: only delete one langtype triple per one graph. Sorry...

Reply:
unless the array is always length == 1

Reply:
It is. Look at [this block of code](https://github.com/lblod/worship-positions-graph-dispatcher-service-loket/blob/954eeea5bb5799edc5b29fce07cdac50046ee64d/lib/deltaProcessing.js#L522). Triples with `termType` `Literal` are filtered from the normal deletion, and are put in stores of size 1, to be deleted one by one in separate queries. I agree that I could have put both workaround in the same place, or structured the code a bit clearer on that front.

Reply:
check